classes/search/cmsfield.php

```diff
@@ -0,0 +1,187 @@
+        $doc->set('modified', $record->timemodified);
+
+        // Check if this document should be considered new.
+        if (isset($options['lastindexedtime']) && ($options['lastindexedtime'] < $record->timecreated)) {
+            // If the document was created after the last index time, it must be new.
+            $doc->set_is_new(true);
+        }
+
+        return $doc;
+    }
+
+    /**
+     * Whether the user can access the document or not.
+     *
+     * @param int $id data id
+     * @return bool
+     */
+    public function check_access($id) {
+        try {
+            $data = $this->get_data($id);
+            $cminfo = $this->get_cm('cms', $data->instanceid, $data->courseid);
+        } catch (\dml_missing_record_exception $ex) {
+            return \core_search\manager::ACCESS_DELETED;
+        } catch (\dml_exception $ex) {
+            return \core_search\manager::ACCESS_DENIED;
+        }
+
+        // Recheck uservisible although it should have already been checked in core_search.
+        if ($cminfo->uservisible === false) {
+            return \core_search\manager::ACCESS_DENIED;
+        }
+
+        $context = \context_module::instance($cminfo->id);
+
+        if (!has_capability('mod/cms:view', $context)) {
+            return \core_search\manager::ACCESS_DENIED;
+        }
+
+        return \core_search\manager::ACCESS_GRANTED;
+    }
+
+    /**
+     * Link to the cms.
+     *
+     * @param \core_search\document $doc
+     * @return \moodle_url
+     */
+    public function get_doc_url(\core_search\document $doc) {
+        $contextmodule = \context::instance_by_id($doc->get('contextid'));
+        $cm = get_coursemodule_from_id('cms', $contextmodule->instanceid, $doc->get('courseid'), true);
+        return new \moodle_url('/course/view.php', ['id' => $doc->get('courseid'), 'section' => $cm->sectionnum]);
+    }
+
+    /**
+     * Link to the cms.
+     *
+     * @param \core_search\document $doc
+     * @return \moodle_url
+     */
+    public function get_context_url(\core_search\document $doc) {
+        $contextmodule = \context::instance_by_id($doc->get('contextid'));
+        return new \moodle_url('/mod/cms/view.php', ['id' => $contextmodule->instanceid]);
+    }
+
+    /**
+     * Returns the specified data from its internal cache.
+     *
+     * @throws \dml_missing_record_exception
+     * @param int $id
+     * @return stdClass
+     */
+    protected function get_data($id) {
+        global $DB;
+        if (empty($this->cmsdata[$id])) {
+            $sql = "SELECT mcd.*, mc.id AS cmsid, mc.course AS courseid
+                      FROM {customfield_data} mcd
+                      JOIN {cms} mc ON mc.id = mcd.instanceid
+                     WHERE mcd.id = :id";
+            $this->cmsdata[$id] = $DB->get_record_sql($sql, ['id' => $id]);
+            if (!$this->cmsdata[$id]) {
+                throw new \dml_missing_record_exception('cms_data');
```

Comment:
Can you please explain why we are throwing dml_missing_record_exception exception here?

Reply:
I updated the `$DB->get_record_sql` to use `MUST_EXIST` option which is using `dml_missing_record_exception` in it.